zkevm-circuits/src/evm_circuit/op_execution/arithmetic.rs

```diff
@@ -0,0 +1,553 @@
+            will_halt: false,
+        },
+        CaseConfig {
+            case: Case::StackUnderflow,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+        CaseConfig {
+            case: Case::OutOfGas,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+    ];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self {
+        let [mut success, stack_underflow, out_of_gas]: [CaseAllocation<F>; 3] =
+            case_allocations.try_into().unwrap();
+        Self {
+            success: AddSuccessAllocation {
+                selector: success.selector,
+                swap: success.cells.pop().unwrap(),
+                a: success.words.pop().unwrap(),
+                b: success.words.pop().unwrap(),
+                c: success.words.pop().unwrap(),
+                carry: success.cells.try_into().unwrap(),
+            },
+            stack_underflow: stack_underflow.selector,
+            out_of_gas: (
+                out_of_gas.selector,
+                out_of_gas.resumption.unwrap().gas_available,
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let OpExecutionState { opcode, .. } = &state_curr;
+
+        let common_polys = vec![
+            (opcode.expr() - OpcodeId::ADD.expr())
+                * (opcode.expr() - OpcodeId::SUB.expr()),
+        ];
+
+        let success = {
+            // interpreter state transition constraints
+            let state_transition_constraints = vec![
+                state_next.global_counter.expr()
+                    - (state_curr.global_counter.expr() + 3.expr()),
+                state_next.program_counter.expr()
+                    - (state_curr.program_counter.expr() + 1.expr()),
+                state_next.stack_pointer.expr()
+                    - (state_curr.stack_pointer.expr() + 1.expr()),
+                state_next.gas_counter.expr()
+                    - (state_curr.gas_counter.expr() + GasCost::FASTEST.expr()),
+            ];
+
+            let AddSuccessAllocation {
+                selector,
+                swap,
+                a,
+                b,
+                c,
+                carry,
+            } = &self.success;
+
+            // swap b and c if it's SUB
+            let no_swap = 1.expr() - swap.expr();
+            let swap_constraints = vec![
+                swap.expr() * no_swap.clone(),
+                swap.expr() * (opcode.expr() - OpcodeId::SUB.expr()),
+                no_swap.clone() * (opcode.expr() - OpcodeId::ADD.expr()),
+            ];
+
+            // add constraints
+            let add_constraints = {
+                let mut constraints = Vec::with_capacity(32);
+                // 256 * carry_out + c
+                let lhs = carry[0].expr() * 256.expr() + c.cells[0].expr();
```

Comment:
Do we need to prove the `carry` belongs to [0, 1]? I wonder if we could find a value other than 1 such that x * 256 = 256 under Fp region.

Reply:
Thanks for catching this! Fixed in [`628bafa`](https://github.com/appliedzkp/zkevm-circuits/commit/628bafa3c9713ccb9e6f353b998722383091db02).